File: livedata-lesson/src/main/java/com/acme/recipes/database/dao/RecipeDao.java
package com.acme.recipes.database.dao;

import android.arch.lifecycle.LiveData;

import com.acme.recipes.database.entity.RecipeEntity;
import com.acme.recipes.database.entity.RecipeEntityFields;
import com.acme.recipes.database.util.RealmLiveData;
import com.acme.recipes.database.util.RealmResultsLiveData;

import javax.annotation.Nonnull;

import io.realm.Realm;
import io.realm.RealmQuery;
import io.realm.RealmResults;

public class RecipeDao extends Dao<RecipeEntity> {

    public RecipeDao(Realm db) {
        super(db);
    }

    public LiveData<RealmResults<RecipeEntity>> findAllAsync() {
        return new RealmResultsLiveData<>(where().findAllAsync());
    }

    public LiveData<RecipeEntity> findByIdAsync(final String id) {
        return new RealmLiveData<>(
                where()
                .equalTo(RecipeEntityFields.ID, id)
                .findFirstAsync());
    }

    private RealmQuery<RecipeEntity> where() {
        return db.where(RecipeEntity.class);
    }

    public void deleteAll() {
        db.executeTransactionAsync(new Realm.Transaction() {
            @Override
            public void execute(@Nonnull Realm bgRealm) {
                bgRealm.delete(RecipeEntity.class);
            }
        });
    }
}

File: app/src/main/java/com/acme/recipes/database/dao/RecipeDao.kt
package com.acme.recipes.database.dao

import android.arch.lifecycle.LiveData
import com.acme.recipes.database.entity.RecipeEntity
import com.acme.recipes.database.entity.RecipeEntityFields
import com.acme.recipes.database.util.asLiveData
import io.realm.Realm
import io.realm.RealmQuery

class RecipeDao(db: Realm) : Dao<RecipeEntity>(db) {

    fun findAllAsync(): LiveData<List<RecipeEntity>> {
        return where().findAllAsync().asLiveData()
    }

    fun findByIdAsync(id: String): LiveData<RecipeEntity> {
        return  where()
                .equalTo(RecipeEntityFields.ID, id)
                .findFirstAsync().asLiveData()
    }

    private fun where(): RealmQuery<RecipeEntity> {
        return db.where(RecipeEntity::class.java)
    }

    fun deleteAll() {
        db.executeTransactionAsync { bgRealm -> bgRealm.delete(RecipeEntity::class.java) }
    }
}
